docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
+
+
+Similar to KIP-612, adapted to Redpanda's thread per core model
+
+## What is being proposed
+
+New configuration properties (For start we choose global config for connection rate):
+- kafka_max_connection_creation_rate (optional integer, default unset)
+- kafka_max_connection_creation_rate_ovverides (string, default "")
+
+
+When a connection is accepted which exceeds one of the configured
+bounds, it is blocking before reading or writing anything to the socket.
+
+
+## Motivation
+
+* Avoid hitting system resource limits (for example CPU)
+* Avoid geting connections storms
+* Allow broker doing useful work instead of handling high connection creation rate
+
+## How
+
+If both configuration properties are unset, then no action is taken: this
+preserves the existing unlimited behavior.
+ 
+We will track connection rate on each shard independently (without cross-core communication).
+ 
+We will store semaphore to count connection in current second. For overrides we will store hash table
+from ip -> semaphore.
+ 
+struct connection_rate {
+   ss::semaphore current_rate
+   ss::lower_clock last_update_time
+}
+ 
+For new connection we will do:
+1. Run wait() for semaphore
+2. Store connection time for signal or not about the end of connection
+3. At the end of connection we should check do we need to signal(1) or not (If second was not changed we need to signal(1) to semaphore)
+
+
+### How to add new tokens to semaphore
+We can run background fiber, which will add kafka_max_connection_creation_rate tokens to semaphore on each second.
+ 
+But we can change algo for connection and do it lazy:
+ 
+At the end of connection we will signal(kafka_max_connection_creation_rate) if the second was changed.
+ 
+Problem: what if we have a blocking connection and nobody signals to it, because other connections are busy.
+We should avoid this situation, and will spawn background fiber if available_units is 0 after lock.
+Background fiber will signal(kafka_max_connection_creation_rate) on each second after spawn if needed (if zero available tokens)
+ 
+Note: each signal(kafka_max_connection_creation_rate) also should update last_update_time
+
+### Handling a configuration change
+
+Changes to the configuration properties affect all outstanding
+connection_rate objects: the `kafka_max_connection_creation_rate` attribute must be recalculated.
+
+### Metrics
```

Comment:
Is it true that connections rate in general it is `kafka_rpc: Currently active connections`? And we are interested only in rate per ip from overrides?